File: 이정선/11월1주차/기지국 설치.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int n, vector<int> stations, int w)
{
    int answer = 0;
    
    vector<int> list;
    
    for(int i = 0; i < stations.size(); i++){
        if(i == 0) {
            if(stations[i] - w - 1 <= 0) continue;
            list.push_back(stations[i] - w - 1);
        } else {
            if(stations[i] - w - (stations[i-1] + w) - 1 <= 0) continue;
            list.push_back(stations[i] - w - (stations[i-1] + w) - 1);
        }
    }
    
    if(stations[stations.size() - 1] != n) {
        if(n - (stations[stations.size() - 1] + w) > 0) {
            list.push_back(n - (stations[stations.size() - 1] + w));
        }

    }
    
    for(int i = 0; i < list.size(); i++){
        answer += (list[i] - 1 )/ (2 * w + 1) + 1;
    }
    
    return answer;
}

File: 송하연/11월 1주차/기지국설치.java
class Solution {
    public int solution(int n, int[] stations, int w) {
        int answer = 0;
        
        for(int i=0; i<stations.length; i++){
            int start=0;
            int end=0;
            
            int minusRange = stations[i]-w;
            int plusRange = stations[i]+w;
     
            if(i==0){
                if(minusRange>1){
                    answer+=install(minusRange-1,w);
                }
                
            }
            else{
                int emptyRange = (stations[i]-w)-(stations[i-1]+w)-1;
                if(emptyRange>0)  answer+=install(emptyRange,w);         
            }

            if(i==stations.length-1){
                if(plusRange<n){
                    answer+=install(n-plusRange,w); 
                } 
            }
        }

        return answer;
    }
    
    //빈 공간에 몇 개 설치 필요한지
    int install(int emptyRange, int w){
        int cover = 2*w+1;
        //남는 부분이 있으면 하나 추가
        int installCnt = emptyRange%cover==0? emptyRange/cover : emptyRange/cover + 1;
        return installCnt;
    }
}
